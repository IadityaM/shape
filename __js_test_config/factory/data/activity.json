{"data":{"id":"1365201","type":"activities","attributes":{"action":null,"created_at":"2019-12-16T23:14:35.973Z","content":null,"actor_id":22548,"target_type":"Collection","target_id":762911,"source_name":""},"relationships":{"actor":{"meta":{"included":false}},"subject_users":{"meta":{"included":false}},"subject_groups":{"meta":{"included":false}},"target":{"meta":{"included":false}}}}}