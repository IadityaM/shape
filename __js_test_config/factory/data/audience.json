{"data":{"id":"104","type":"audiences","attributes":{"name":"Natural Armor","global_default":null,"age_list":[],"children_age_list":[],"country_list":[],"education_level_list":[],"gender_list":[],"adopter_type_list":[],"interest_list":[],"publication_list":[],"min_price_per_response":4.0,"order":null,"global":true}}}