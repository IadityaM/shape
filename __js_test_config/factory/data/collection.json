{"data":{"id":"762903","type":"collections","attributes":{"class_type":"Collection","name":"Extended","created_at":"2019-12-16T23:14:33.643Z","updated_at":"2019-12-16T23:14:33.643Z","master_template":false,"template_id":null,"submission_box_type":null,"submission_box_id":null,"submission_template_id":null,"test_status":null,"collection_to_test_id":null,"hide_submissions":false,"submissions_enabled":true,"anyone_can_view":false,"anyone_can_join":false,"cover_type":"cover_type_default","archived":false,"unresolved_count":0,"last_unresolved_comment_id":null,"test_show_media":true,"idea_id":null,"search_term":null,"organization_id":"5352","joinable_group_id":null,"default_group_id":null,"system_required":false,"tag_list":[],"inherited_tag_list":[],"cover":{"text":"","hardcoded_subtitle":"","subtitle_hidden":false},"test_scores":{},"breadcrumb":[{"type":"collections","id":"762903","name":"Extended","can_edit":false}],"processing_status":null,"collection_card_count":0,"can_view":false,"can_edit":false,"can_edit_content":false,"is_private":false,"is_org_template_collection":false,"is_profile_template":false,"is_profile_collection":false,"pinned_and_locked":false,"is_submission_box_template":false,"is_submission_box_template_test":false,"is_inside_hidden_submission_box":false,"is_inside_a_submission":false,"is_subtemplate_or_instance":false,"show_language_selector":false,"template_num_instances":0,"awaiting_updates":null,"num_survey_responses":0,"max_row_index":null,"max_col_index":null,"frontend_url":"http://localhost/","common_viewable":null,"is_test_locked":null,"has_link_sharing":null,"is_restorable":false,"serializer":"SerializableCollection"},"relationships":{"roles":{"meta":{"included":false}},"collection_cover_items":{"meta":{"included":false}},"parent_collection_card":{"meta":{"included":false}},"parent":{"meta":{"included":false}},"live_test_collection":{"meta":{"included":false}},"submissions_collection":{"meta":{"included":false}},"submission_template":{"meta":{"included":false}},"collection_to_test":{"meta":{"included":false}},"organization":{"meta":{"included":false}},"created_by":{"meta":{"included":false}},"test_audiences":{"meta":{"included":false}},"test_results_collection":{"meta":{"included":false}},"collection_filters":{"meta":{"included":false}},"restorable_parent":{"meta":{"included":false}}}}}