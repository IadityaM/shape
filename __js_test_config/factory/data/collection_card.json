{"data":{"id":"2877260","type":"collection_cards","attributes":{"class_type":"CollectionCard::Primary","order":0,"width":1,"height":1,"parent_id":762904,"pinned":false,"image_contain":false,"card_question_type":null,"is_cover":false,"filter":"transparent_gray","hidden":false,"show_replace":true,"updated_at":"2019-12-16T23:14:34.478Z","col":1,"row":1,"section_type":null,"idea_id":null,"link":false,"pinned_and_locked":false,"can_edit_parent":false,"is_master_template_card":false},"relationships":{"item":{"meta":{"included":false}},"collection":{"meta":{"included":false}},"parent":{"meta":{"included":false}},"record":{"meta":{"included":false}},"templated_from":{"meta":{"included":false}}}}}