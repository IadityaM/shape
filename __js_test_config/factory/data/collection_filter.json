{"data":{"id":"18","type":"collection_filters","attributes":{"text":"orchid","filter_type":"search","selected":false}}}