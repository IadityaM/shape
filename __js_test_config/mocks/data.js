export const fakeTextItemAttrs = {
  id: 1,
  type: 'Item::TextItem',
  name: 'My Cool Item',
  text_data: 'This is the content for the item and it contains multiple sentences. Like this one.',
  breadcrumb: [['collections', 1, 'Some collection'], ['items', 1, 'my item']],
  parentPath: '/',
}
export const fakeTextItem = {
  ...fakeTextItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
  toJS: jest.fn().mockReturnValue(fakeTextItemAttrs),
}
export const fakeVideoItemAttrs = {
  id: 2,
  type: 'Item::VideoItem',
  name: 'Crazy Honey Badger',
  url: 'https://www.youtube.com/watch?v=4r7wHMg5Yjg'
}
export const fakeVideoItem = {
  ...fakeVideoItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeImageItemAttrs = {
  id: 3,
  type: 'Item::ImageItem',
  name: 'Earth from Space',
  filestack_file: {
    id: 1,
    url: 'https://www.nasa.gov/sites/default/files/styles/full_width_feature/public/thumbnails/image/iss052e023801_0.jpg',
  }
}
export const fakeImageItem = {
  ...fakeImageItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeTextItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeItemCard = {
  id: 10,
  order: 0,
  height: 1,
  width: 1,
  record: fakeTextItem,
  item: fakeTextItem,
}
export const fakeCollectionCard = {
  id: 11,
  order: 1,
  height: 1,
  width: 1,
  record: fakeCollection,
  item: fakeCollection,
  reference: false,
}
export const fakeCollection = {
  id: 1,
  name: 'My Workspace X',
  type: 'Collection',
  breadcrumb: [],
  collection_cards: [
    fakeItemCard, fakeItemCard, fakeItemCard
  ]
}
export const fakePosition = {
  xPos: 0,
  yPos: 0,
  height: 1,
  width: 1,
}
