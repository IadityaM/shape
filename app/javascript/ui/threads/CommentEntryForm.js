import PropTypes from 'prop-types'
import { toJS, observable, action, runInAction } from 'mobx'
import { observer, PropTypes as MobxPropTypes } from 'mobx-react'
import { EditorState, ContentState, convertToRaw } from 'draft-js'

import ReturnArrowIcon from '~/ui/icons/ReturnArrowIcon'
import { CommentForm } from '~/ui/global/styled/forms'
import CommentInput from './CommentInput'

@observer
class CommentEntryForm extends React.Component {
  @observable commentData = {
    message: '',
    draftjs_data: {},
  }
  @observable suggestionsOpen = false
  @observable updating = false
  // @observable editorState = EditorState.createEmpty()
  state = {
    editorState: EditorState.createEmpty()
  }

  componentDidMount() {
    this.focusTextArea(this.props.expanded)
  }

  componentWillReceiveProps({ expanded }) {
    this.focusTextArea(expanded)
    // NOTE: maybe preferred to leave written + unsent messages in the comment box?
    // if (!expanded) this.resetEditorState()
  }

  componentWillUnmount() {
    this.editor = null
  }

  focusTextArea = (expanded) => {
    // NOTE: draft-js-plugins need timeout, even with 0 delay, see:
    // https://github.com/draft-js-plugins/draft-js-plugins/issues/800#issuecomment-315950836
    setTimeout(() => {
      if (!expanded || !this.editor) return
      this.editor.focus()
    })
  }

  @action handleInputChange = (editorState) => {
    if (this.updating) return
    const content = editorState.getCurrentContent()
    const message = content.getPlainText()
    this.commentData.message = message
    this.commentData.draftjs_data = convertToRaw(content)
    // this.editorState = editorState
    this.setState({
      editorState
    })
  }

  setEditor = (editor, { unset = false } = {}) => {
    if (unset) {
      this.editor = null
      return
    }
    if (this.editor) return
    this.editor = editor
    this.focusTextArea(this.props.expanded)
  }

  resetEditorState() {
    // this.editorState = EditorState.push(this.state.editorState, ContentState.createFromText(''))
    this.setState({
      editorState: EditorState.push(this.state.editorState, ContentState.createFromText(''))
    })
  }

  @action handleOpenSuggestions = () => {
    this.suggestionsOpen = true
  }

  @action handleCloseSuggestions = () => {
    this.suggestionsOpen = false
  }

  handleReturn = (e) => {
    if (!e.shiftKey && !this.suggestionsOpen) {
      // submit message
      this.handleSubmit(e)
      return 'handled'
    }
    return 'not-handled'
  }

  handleSubmit = (e) => {
    e.preventDefault()
    const rawData = toJS(this.commentData)
    // don't allow submit of empty comment
    if (!rawData.message) return
    const { thread } = this.props
    thread.API_saveComment(rawData).then(() => {
      this.props.afterSubmit()
      this.updating = false
    })
    runInAction(() => {
      this.commentData.message = ''
      this.commentData.draftjs_data = {}
      this.updating = true
    })
    this.resetEditorState()
  }

  render() {
    const { expanded } = this.props
    if (!expanded) return ''

    return (
      <CommentForm className="reply" onSubmit={this.handleSubmit}>
        <div className="textarea-input">
          <CommentInput
            editorState={this.state.editorState}
            onChange={this.handleInputChange}
            onOpenSuggestions={this.handleOpenSuggestions}
            onCloseSuggestions={this.handleCloseSuggestions}
            handleReturn={this.handleReturn}
            setEditor={this.setEditor}
          />
        </div>
        <button>
          <ReturnArrowIcon />
        </button>
      </CommentForm>
    )
  }
}

CommentEntryForm.propTypes = {
  expanded: PropTypes.bool.isRequired,
  afterSubmit: PropTypes.func.isRequired,
  thread: MobxPropTypes.objectOrObservableObject.isRequired,
}

export default CommentEntryForm
