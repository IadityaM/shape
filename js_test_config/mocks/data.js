export const fakeItemAttrs = {
  id: 1,
  name: 'My Cool Item',
}
export const fakeItem = {
  ...fakeItemAttrs,
  rawAttributes: jest.fn().mockReturnValue(fakeItemAttrs),
  getRecordType: jest.fn().mockReturnValue('items'),
}
export const fakeCard = {
  id: 10,
  order: 0,
  height: 1,
  width: 1,
  record: fakeItem,
  item: fakeItem,
}
export const fakeCollection = {
  id: 1,
  name: 'My Workspace X',
  breadcrumb: [],
  collection_cards: [
    fakeCard, fakeCard, fakeCard
  ]
}
export const fakePosition = {
  xPos: 0,
  yPos: 0,
  height: 1,
  width: 1,
}
