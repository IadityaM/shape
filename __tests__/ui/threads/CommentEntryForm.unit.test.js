import CommentEntryForm from '~/ui/threads/CommentEntryForm'
import { fakeThread } from '#/mocks/data'

let wrapper, props
describe('CommentEntryForm', () => {
  beforeEach(() => {
    props = {
      expanded: true,
      afterSubmit: jest.fn(),
      thread: fakeThread,
    }
    wrapper = shallow(<CommentEntryForm {...props} />)
    fakeThread.API_saveComment.mockClear()
  })

  describe('when not expanded', () => {
    beforeEach(() => {
      wrapper.setProps({ ...props, expanded: false })
    })

    it('does not render the CommentInput box', () => {
      expect(wrapper.find('CommentInput').exists()).toBeFalsy()
    })
  })

  describe('when expanded', () => {
    beforeEach(() => {
      wrapper.setProps({ ...props, expanded: true })
    })

    it('renders the CommentInput box', () => {
      expect(wrapper.find('CommentInput').exists()).toBeTruthy()
    })
  })
})
