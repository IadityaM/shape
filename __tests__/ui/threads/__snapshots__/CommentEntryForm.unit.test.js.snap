// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentEntryForm renders snapshot 1`] = `
<CommentForm
  onSubmit={[Function]}
>
  <StyledCommentInputWrapper
    className="textarea-input"
    replying={false}
  >
    <CommentInput
      editorState={
        EditorState {
          "_immutable": Immutable.Record {
            "allowUndo": true,
            "currentContent": Immutable.Record {
              "entityMap": Object {
                "__add": [Function],
                "__create": [Function],
                "__get": [Function],
                "__getLastCreatedEntityKey": [Function],
                "__mergeData": [Function],
                "__replaceData": [Function],
                "add": [Function],
                "create": [Function],
                "get": [Function],
                "getLastCreatedEntityKey": [Function],
                "mergeData": [Function],
                "replaceData": [Function],
              },
              "blockMap": Immutable.OrderedMap {
                "18kt7": Immutable.Record {
                  "key": "18kt7",
                  "type": "unstyled",
                  "text": "",
                  "characterList": Immutable.List [],
                  "depth": 0,
                  "data": Immutable.Map {},
                },
              },
              "selectionBefore": Immutable.Record {
                "anchorKey": "18kt7",
                "anchorOffset": 0,
                "focusKey": "18kt7",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
              "selectionAfter": Immutable.Record {
                "anchorKey": "18kt7",
                "anchorOffset": 0,
                "focusKey": "18kt7",
                "focusOffset": 0,
                "isBackward": false,
                "hasFocus": false,
              },
            },
            "decorator": null,
            "directionMap": Immutable.OrderedMap {
              "18kt7": "LTR",
            },
            "forceSelection": false,
            "inCompositionMode": false,
            "inlineStyleOverride": null,
            "lastChangeType": null,
            "nativelyRenderedContent": null,
            "redoStack": Immutable.Stack [],
            "selection": Immutable.Record {
              "anchorKey": "18kt7",
              "anchorOffset": 0,
              "focusKey": "18kt7",
              "focusOffset": 0,
              "isBackward": false,
              "hasFocus": false,
            },
            "treeMap": Immutable.OrderedMap {
              "18kt7": Immutable.List [
                Immutable.Record {
                  "start": 0,
                  "end": 0,
                  "decoratorKey": null,
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "start": 0,
                      "end": 0,
                    },
                  ],
                },
              ],
            },
            "undoStack": Immutable.Stack [],
          },
        }
      }
      handleSubmit={[Function]}
      onChange={[Function]}
      setEditor={[Function]}
    />
  </StyledCommentInputWrapper>
  <styled.button
    focused={false}
  >
    <ReturnArrowIcon />
  </styled.button>
</CommentForm>
`;
